File: sdk/python/hitechcloud/resources/cloud_vm.py
"""Cloud Virtual Machine resource (9 endpoints)"""

from .base import BaseResource

class CloudVmResource(BaseResource):
    """Cloud Virtual Machine: upgrade, resources, templates, ISO, boot, PXE"""

    def get_upgrade_options(self, service_id: int) -> dict:
        """GET /api/service/{id}/upgrade - List resource upgrade options"""
        return self._http.get(f"/api/service/{service_id}/upgrade")

    def request_upgrade(self, service_id: int, **kwargs) -> dict:
        """POST /api/service/{id}/upgrade - Estimate or request upgrade"""
        return self._http.post(f"/api/service/{service_id}/upgrade", kwargs)

    def get_resources(self, service_id: int) -> dict:
        """GET /api/service/{id}/resources - Show available and used resources"""
        return self._http.get(f"/api/service/{service_id}/resources")

    def list_rebuild_templates(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/templates/{vmid} - List rebuild templates"""
        return self._http.get(f"/api/service/{service_id}/templates/{vmid}")

    def get_memory_usage(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/usage/memory - Memory usage graph"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/usage/memory")

    def list_iso_images(self, service_id: int, vmid: int) -> dict:
        """GET /api/service/{id}/vms/{vmid}/images - List ISO images for VM"""
        return self._http.get(f"/api/service/{service_id}/vms/{vmid}/images")

    def mount_iso(self, service_id: int, vmid: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms/{vmid}/images - Mount/Unmount ISO image"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/images", kwargs)

    def set_boot_order(self, service_id: int, vmid: int, **kwargs) -> dict:
        """POST /api/service/{id}/vms/{vmid}/boot - Change boot order"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/boot", kwargs)

    def enable_pxe(self, service_id: int, vmid: int) -> dict:
        """POST /api/service/{id}/vms/{vmid}/tuntap - Enable PXE"""
        return self._http.post(f"/api/service/{service_id}/vms/{vmid}/tuntap")

File: sdk/ruby/lib/hitechcloud/resources/cloud_vm.rb
module HiTechCloud
  module Resources
    class CloudVm < Base

        # Get upgrade options
        def upgrade(id)
          @http.get("/api/service/%{id}/upgrade" % { id: id })
        end

        # Upgrade
        def do_upgrade(id, data)
          @http.post("/api/service/%{id}/upgrade" % { id: id }, data)
        end

        # Get resources
        def resources(id)
          @http.get("/api/service/%{id}/resources" % { id: id })
        end

        # Get template
        def template(id, vmid)
          @http.get("/api/service/%{id}/templates/%{vmid}" % { id: id, vmid: vmid })
        end

        # Get memory usage
        def memory_usage(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/usage/memory" % { id: id, vmid: vmid })
        end

        # Get VM images
        def vm_images(id, vmid)
          @http.get("/api/service/%{id}/vms/%{vmid}/images" % { id: id, vmid: vmid })
        end

        # Create VM image
        def create_vm_image(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/images" % { id: id, vmid: vmid }, data)
        end

        # Set boot order
        def boot_order(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/boot" % { id: id, vmid: vmid }, data)
        end

        # Toggle TUN/TAP
        def tuntap(id, vmid, data)
          @http.post("/api/service/%{id}/vms/%{vmid}/tuntap" % { id: id, vmid: vmid }, data)
      end
    end
  end
end
